include/npcomp/Dialect/Basicpy/IR/BasicpyDialect.td

```diff
@@ -96,6 +96,36 @@ def Basicpy_UnknownType : DialectType<Basicpy_Dialect,
   }];
 }
 
+def Basicpy_ListType : DialectType<Basicpy_Dialect,
+    CPred<"$_self.isa<::mlir::NPCOMP::Basicpy::ListType>()">,
+    "List type"> {
+  let typeDescription = [{
```

Comment:
In principle, a Python list can only hold python objects. However, it seems that here we allow any MLIR type (not requiring a "HasPythonBinding" trait or such shenanigans). Is there some way you can massage this description (or maybe do it at the dialect-doc level) about that detail? Maybe "We assume that any MLIR type can in principle be represented as a Python object, so we allow any MLIR type to be held inside the "container" types such as dict/list/tuple. We anticipate that during lowering, the compiler will need to make decisions about how to lower the list/tuple/dict container types to more primitive types, along with how to lower the contained types, which will dictate the ultimately valid contained type possibilities. As such, we don't impose any formal requirement/trait on the types that are contained in these containers."

Reply:
I had a TODO to tighten the constraints and just expanded that in the direction you mention. At this point, the lack of a more consistent constraint should be interpreted as "not designed yet" versus intentional.